sympy/parsing/mathematica.py

```diff
@@ -1,67 +1,133 @@
 from __future__ import print_function, division
 
-from re import match
+import re
 from sympy import sympify
 
 
 def mathematica(s):
     return sympify(parse(s))
 
 
-def parse(s):
-    s = s.strip()
-
-    # Begin rules
-    rules = (
-        # Arithmetic operation between a constant and a function
-        (r"\A(\d+)([*/+-^])(\w+\[[^\]]+[^\[]*\])\Z",
-        lambda m: m.group(
-            1) + translateFunction(m.group(2)) + parse(m.group(3))),
-
-        # Arithmetic operation between two functions
-        (r"\A(\w+\[[^\]]+[^\[]*\])([*/+-^])(\w+\[[^\]]+[^\[]*\])\Z",
-        lambda m: parse(m.group(1)) + translateFunction(
-            m.group(2)) + parse(m.group(3))),
-
-        (r"\A([+-]?)(\w+)\[([^\]]+[^\[]*)\]\Z",  # Function call
-        lambda m: m.group(1) + translateFunction(
-            m.group(2)) + "(" + parse(m.group(3)) + ")"),
-
-        (r"\((.+)\)\((.+)\)",  # Parenthesized implied multiplication
-        lambda m: "(" + parse(m.group(1)) + ")*(" + parse(m.group(2)) + ")"),
-
-        (r"\A\((.+)\)\Z",  # Parenthesized expression
-        lambda m: "(" + parse(m.group(1)) + ")"),
-
-        (r"\A(.*[\w\.])\((.+)\)\Z",  # Implied multiplication - a(b)
-        lambda m: parse(m.group(1)) + "*(" + parse(m.group(2)) + ")"),
-
-        (r"\A\((.+)\)([\w\.].*)\Z",  # Implied multiplication - (a)b
-        lambda m: "(" + parse(m.group(1)) + ")*" + parse(m.group(2))),
-
-        (r"\A(-? *[\d\.]+)([a-zA-Z].*)\Z",  # Implied multiplication - 2a
-        lambda m: parse(m.group(1)) + "*" + parse(m.group(2))),
-
-        (r"\A([^=]+)([\^\-\*/\+=]=?)(.+)\Z",  # Infix operator
-        lambda m: parse(m.group(1)) + translateOperator(m.group(2)) + parse(m.group(3))))
-    # End rules
-
-    for rule, action in rules:
-        m = match(rule, s)
-        if m:
-            return action(m)
-
-    return s
-
-
-def translateFunction(s):
-    if s.startswith("Arc"):
-        return "a" + s[3:]
-    return s.lower()
-
-
-def translateOperator(s):
-    dictionary = {'^': '**'}
-    if s in dictionary:
-        return dictionary[s]
-    return s
+class _Parse(object):
+    '''An instance of this class converts basic Mathematica expression to
+    Python.'''
+
+    replaces1 = (
+        (' ', ''),
+        ('^', '**'),
+    )
+
+    replaces2 = (
+        ('[', '('),
+        (']', ')'),
+    )
+
+    rules1 = (
+        # 'Arc' to 'a'
+        (r'''
+                (               # group1
+                .*              # any characters
+                [^a-zA-Z]       # a single character except alphabet
+                )
+                Arc             # detect 'Arc'
```

Comment:
```
In[1]:= ArcSin[1/2]                                                             

        Pi
Out[1]= --
        6

In[2]:= Arcsin[1/2]                                                             

               1
Out[2]= Arcsin[-]
               2
```
Are you sure it is parsing correctly?

Reply:
I mean, `Arcsin` is a wrong spelling, it shouldn't be parsed to `asin`.

Reply:
Thank you for pointing it out.
I supposed that all input were correct expression in Mathematica.
The current mathematica.py also parses 'Arcsin[1/2]' to 'pi/6'.

Your point is that an input-check function should be introduced, right?

Reply:
I'm in favor of committing this as is and letting that issue be a separate PR. Any objections?

Reply:
No objections.

Reply:
Thanks for kind reviewing, everybody.

Reply:
I'd like to start working on checking the input expressions if nobody else has started on that.

Reply:
Thank you, kengo! But, soon I will send new pull request for making further rigid pursing.
So just wait a little bit, please.
And I am sorry if you already started checking.